### PR TITLE
Project resources

### PR DESCRIPTION
**Pull request by kvikshaug**
- Adds CRUD resources for projects.
- Adds jwt verification module, since IAM now accepts requests requiring authorization. The implementation is almost identical to other services with a `jwt_required` decorator and `jwt_require_claim` function.
- `usr` key is added to JWT claims in order to identify the user for a given token
- Note that org/team relationships are not yet implemented; users are simply given admin role to any project they create, and the role can not be modified.

### CODE REVIEW

File: src/iam/jwt.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    @app.before_request
+    def decode_jwt():
+        if 'Authorization' not in request.headers:
+            logger.debug("No JWT provided")
+            g.jwt_valid = False
+            g.jwt_claims = {'prj': {}}
+            return
+
+        auth = request.headers['Authorization']
+        if not auth.startswith('Bearer '):
```
Comment (Midnighter) [inline src/iam/jwt.py]: Maybe add a log statement here for parity with the above condition?

File: src/iam/jwt.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            g.jwt_valid = False
+            g.jwt_claims = {'prj': {}}
+            return
+
+        try:
+            _, token = auth.split(' ', 1)
```
Comment (Midnighter) [inline src/iam/jwt.py]:
```suggestion
            token = auth.split(' ', 1)[1]
```
Could also use `auth.split(maxsplit=1)` for splitting on any white space.

File: src/iam/jwt.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                token, app.config['RSA_PUBLIC_KEY'], app.config['ALGORITHM'])
+            # JSON object names can only be strings. Map project ids to ints for
+            # easier handling
+            g.jwt_claims['prj'] = {
+                int(key): value
+                for key, value in g.jwt_claims['prj'].items()
+            }
+            g.jwt_valid = True
+            logger.debug(f"JWT claims accepted: {g.jwt_claims}")
+        except (jwt.JWTError, jwt.ExpiredSignatureError,
```
Comment (Midnighter) [inline src/iam/jwt.py]: If the above `auth.split` fails to unpack this will yield a `ValueError` or with my suggestion an `IndexError`. That should probably be caught here?

File: src/iam/jwt.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    :param required_level: The required access level (admin, write or read)
+    :return: None
+    """
+    ACCESS_LEVELS = {
+        'admin': 3,
+        'write': 2,
+        'read': 1,
+    }
+
+    if required_level not in ACCESS_LEVELS.keys():
```
Comment (Midnighter) [inline src/iam/jwt.py]:
```suggestion
    if required_level not in ACCESS_LEVELS:
```

Checking for a key in a dict is O(1) with this simpler syntax.

File: src/iam/jwt.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            g.jwt_claims = jwt.decode(
+                token, app.config['RSA_PUBLIC_KEY'], app.config['ALGORITHM'])
+            # JSON object names can only be strings. Map project ids to ints for
+            # easier handling
+            g.jwt_claims['prj'] = {
+                int(key): value
+                for key, value in g.jwt_claims['prj'].items()
+            }
+            g.jwt_valid = True
+            logger.debug(f"JWT claims accepted: {g.jwt_claims}")
```
Comment (Midnighter) [inline src/iam/jwt.py]:
```suggestion
            logger.debug("JWT claims accepted: %r", g.jwt_claims)
```

File: src/iam/jwt.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    ACCESS_LEVELS = {
+        'admin': 3,
+        'write': 2,
+        'read': 1,
+    }
+
+    if required_level not in ACCESS_LEVELS.keys():
+        raise ValueError(f"Invalid claim level '{required_level}'")
+
+    logger.debug(f"Looking for '{required_level}' access to project "
```
Comment (Midnighter) [inline src/iam/jwt.py]: As above I'd suggest to use different formatting.

File: src/iam/resources.py [hunk 7]
```diff
@@ -118,14 +121,73 @@ def post(self, refresh_token):
...
+        db.session.add(user_project)
+        db.session.commit()
+        return {'project_id': project.id}, 201
+
+
+@doc(description="List projects")
+class ProjectResource(MethodResource):
+    @marshal_with(ProjectResponseSchema(), code=200)
+    def get(self, project_id):
+        try:
```
Comment (Midnighter) [inline src/iam/resources.py]: Could be refactored into a static utility method.

File: src/iam/jwt.py (hunk 3 above)
Comment (kvikshaug) [reply src/iam/jwt.py]: This actually cannot be the case because of [the check on L39](https://github.com/DD-DeCaF/iam/pull/14/files#diff-546df41352a3e0c0653b4d85eaf38494R39) guarantees a space char in the value. That's not really obvious so I will add a comment.

File: src/iam/jwt.py (hunk 4 above)
Comment (kvikshaug) [reply src/iam/jwt.py]: Ok, I realize it is the more pythonic way of checking for a key in a dict and I will try to update my habits :) But just to be clear, in Python 3, using `.keys()` to check on a dictview is still O(1).

File: src/iam/resources.py (hunk 7 above)
Comment (kvikshaug) [reply src/iam/resources.py]: My humble opinion on this is that duplication is cheaper than the wrong abstraction, and this type of generalization is more harmful than helpful in the long run.

File: src/iam/jwt.py (hunk 5 above)
Comment (kvikshaug) [reply src/iam/jwt.py]: I think the performance hit is completely negligible and I value readability (in this case, consistent formatting style) higher.

File: src/iam/resources.py (hunk 7 above)
Comment (Midnighter) [reply src/iam/resources.py]: I don't understand yet what you mean. It's not only about de-duplicating code it's much more about putting it into logical units. Refactoring this piece of code into a method means that I only need to figure out the code once and can then see three times, "oh yeah, it's that same logic again that I already understand". So in my opinion it greatly reduces the effort needed to understand what the class does. As it stands I have to read the code in each method just to realize it does the same thing. I'm okay with agreeing to disagree but I really don't get why you consider it harmful.

Comment (kvikshaug) [reply src/iam/resources.py]: I'll try to explain my reasoning; let me emphasize that I only wish to elaborate about my thinking and the general principle, and not to bikeshed about this specific case :slightly_smiling_face: 

> it's much more about putting it into logical units

This is not a logical unit that makes sense to me. The logic it encapsulates is so straight forward that it really only "wraps" a few APIs that are already designed to perform this logic. The method would still need to throw an exception, and the resource would still need to handle it. So you'd end up with close to the same number of lines, but a new custom API.

The callers could also eventually develop different requirements from the encapsulated logic. For example, the `GET` request might want to query for private or public data (project with NULL id), while `PUT` does not allow updates to public data. (Admittedly this is not the case in the IAM database, but that _is_ the case for f.ex. the model-storage - compare [GET](https://github.com/DD-DeCaF/model-storage/blob/devel/src/model_storage/resources.py#L98) with [PUT](https://github.com/DD-DeCaF/model-storage/blob/devel/src/model_storage/resources.py#L112).) Then you need to add a parameter to the utility method and a code path branch which increases the cyclomatic complexity. To me, it is much more straight forward to simply use the SA API in the resource directly.

(The above point is explained well in [this blog post](https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction))

> Refactoring this piece of code into a method means that I only need to figure out the code once and can then see three times, "oh yeah, it's that same logic again that I already understand". So in my opinion it greatly reduces the effort needed to understand what the class does.

I would actually claim the opposite, because a developer will more likely be familiar with the API surface of the libraries we are using, than with custom abstractions in our classes. Looking at a try/catch block with an SA query is *more* immediately obvious to me, than seeing some method call to another place in the code which I don't know or trust in the same way. It's an added layer of indirection, so I need to context switch, go understand the new method, and yes, then I might get the benefit of immediately understanding the method call at another point in the code.

tl;dr:

* There is not enough encapsulated logic that there really are benefits to share it
* The little logic there is, is not unique enough and may develop different needs for different callers

Comment (Midnighter) [reply src/iam/resources.py]: Thanks for the link to the blog post. Taking the summary paragraph at the end, the argument seems to be to break existing abstractions and re-introduce seeming duplication. So I absolutely agree that at the point when, for example, GET and PUT do different things, it should not be pressed into the utility method but broken apart again. The blog post does not say that creating the abstraction in the first place is wrong, though, and it even recommends creating new abstractions for your new situation.

> The moral of this story? Don't get trapped by the sunk cost fallacy. If you find yourself passing parameters and adding conditional paths through shared code, the abstraction is incorrect. It may have been right to begin with, but that day has passed. Once an abstraction is proved wrong the best strategy is to re-introduce duplication and let it show you what's right. Although it occasionally makes sense to accumulate a few conditionals to gain insight into what's going on, you'll suffer less pain if you abandon the wrong abstraction sooner rather than later.

So two more points to add:

1. As described in the blog post, a future programmer might look at the method and feel compelled to stick to the abstraction and modify it rather than duplicating the code as discussed above. So yes, introducing the method might lead future developers down the wrong path. This could be avoided with a comment maybe. So I prefer documenting current design to assuming responsibility for future errors.
2. Abstracting logic also helps with unit testing. I can test model loading logic in isolation without having to invoke the API of the client.
